// Copyright 2023 Ant Group Co., Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

syntax = "proto3";

package scql.engine;

import "api/engine.proto";
import "pkg/audit/common.proto";

message AuditLog {
  audit.pb.AuditHeader header = 1;
  AuditBody body = 2;
}

message AuditBody {
  oneof body {
    audit.pb.UncategorizedEvent uncategorized = 1;
    RunExecPlanEvent run_plan = 3;
    CreateSessionEvent create_session = 4;
    StopJobEvent stop_job = 5;

    JoinNodeDetail join_detail = 6;
    InNodeDetail in_detail = 7;
    SqlNodeDetail sql_detail = 8;
    DumpFileNodeDetail dump_detail = 9;
    PublishNodeDetail publish_detail = 10;
  }
}

message RunExecPlanEvent {
  string source_ip = 1;
  int64 cost_time = 2;  // unit: ms
  int64 affected_rows = 3;
  repeated audit.pb.NodeInfo node_list = 4;
}

message CreateSessionEvent {
  string source_ip = 1;
  bool sync = 2;
  scql.pb.JobStartParams parameters = 3;
}

message StopJobEvent {
  string reason = 1;
  string source_ip = 2;
}

message JoinNodeDetail {
  string node_name = 1;
  string self_party = 2;
  int64 self_rank = 3;
  int64 self_size = 4;
  int64 peer_size = 5;
  int64 result_size = 6;
  int64 cost_time = 7;  // unit: ms
  repeated string party_codes = 8;
}

message InNodeDetail {
  string node_name = 1;
  string self_party = 2;
  string reveal_to = 3;

  int64 self_size = 4;    // only works when reveal to myself
  int64 peer_size = 5;    // only works when reveal to myself
  int64 result_size = 6;  // only works when reveal to myself

  repeated string party_codes = 7;
  int64 cost_time = 8;  // unit: ms
}

message SqlNodeDetail {
  string node_name = 1;
  string query = 2;
  int64 num_rows = 3;
  int64 num_columns = 4;
  int64 cost_time = 5;  // unit: ms
}

message PublishNodeDetail {
  string node_name = 1;
  int64 num_rows = 2;
  int64 cost_time = 3;  // unit: ms
}

message DumpFileNodeDetail {
  string node_name = 1;
  string file_path = 2;
  int64 affected_rows = 3;
  int64 cost_time = 4;  // unit: ms
}
